// @mysten/wallet-standard/dist/index.mjs
import { signAndExecuteTransaction, signTransaction } from "./wallet.mjs";
import { SuiSignMessage } from "./features/suiSignMessage.mjs";
import { SuiSignTransactionBlock } from "./features/suiSignTransactionBlock.mjs";
import { SuiSignTransaction } from "./features/suiSignTransaction.mjs";
import { SuiSignAndExecuteTransactionBlock } from "./features/suiSignAndExecuteTransactionBlock.mjs";
import { SuiSignAndExecuteTransaction } from "./features/suiSignAndExecuteTransaction.mjs";
import { SuiSignPersonalMessage } from "./features/suiSignPersonalMessage.mjs";
import { SuiGetCapabilities } from "./features/suiGetCapabilities.mjs";
import { isWalletWithRequiredFeatureSet } from "./detect.mjs";
import { SUI_CHAINS, SUI_DEVNET_CHAIN, SUI_LOCALNET_CHAIN, SUI_MAINNET_CHAIN, SUI_TESTNET_CHAIN, isSuiChain } from "./chains.mjs";

export * from "@wallet-standard/core"

export { SUI_CHAINS, SUI_DEVNET_CHAIN, SUI_LOCALNET_CHAIN, SUI_MAINNET_CHAIN, SUI_TESTNET_CHAIN, SuiGetCapabilities, SuiSignAndExecuteTransaction, SuiSignAndExecuteTransactionBlock, SuiSignMessage, SuiSignPersonalMessage, SuiSignTransaction, SuiSignTransactionBlock, isSuiChain, isWalletWithRequiredFeatureSet, signAndExecuteTransaction, signTransaction };
