// @mysten/sui/dist/grpc/proto/sui/rpc/v2/jwk.mjs
import { MessageType } from "@protobuf-ts/runtime";

//#region src/grpc/proto/sui/rpc/v2/jwk.ts
var JwkId$Type = class extends MessageType {
	constructor() {
		super("sui.rpc.v2.JwkId", [{
			no: 1,
			name: "iss",
			kind: "scalar",
			opt: true,
			T: 9
		}, {
			no: 2,
			name: "kid",
			kind: "scalar",
			opt: true,
			T: 9
		}]);
	}
};
/**
* @generated MessageType for protobuf message sui.rpc.v2.JwkId
*/
const JwkId = new JwkId$Type();
var Jwk$Type = class extends MessageType {
	constructor() {
		super("sui.rpc.v2.Jwk", [
			{
				no: 1,
				name: "kty",
				kind: "scalar",
				opt: true,
				T: 9
			},
			{
				no: 2,
				name: "e",
				kind: "scalar",
				opt: true,
				T: 9
			},
			{
				no: 3,
				name: "n",
				kind: "scalar",
				opt: true,
				T: 9
			},
			{
				no: 4,
				name: "alg",
				kind: "scalar",
				opt: true,
				T: 9
			}
		]);
	}
};
/**
* @generated MessageType for protobuf message sui.rpc.v2.Jwk
*/
const Jwk = new Jwk$Type();

//#endregion
export { Jwk, JwkId };
