// @mysten/sui/dist/grpc/proto/sui/rpc/v2/move_package.mjs
import { MessageType } from "@protobuf-ts/runtime";

//#region src/grpc/proto/sui/rpc/v2/move_package.ts
/**
* @generated from protobuf enum sui.rpc.v2.DatatypeDescriptor.DatatypeKind
*/
let DatatypeDescriptor_DatatypeKind = /* @__PURE__ */ function(DatatypeDescriptor_DatatypeKind$1) {
	/**
	* @generated from protobuf enum value: DATATYPE_KIND_UNKNOWN = 0;
	*/
	DatatypeDescriptor_DatatypeKind$1[DatatypeDescriptor_DatatypeKind$1["DATATYPE_KIND_UNKNOWN"] = 0] = "DATATYPE_KIND_UNKNOWN";
	/**
	* @generated from protobuf enum value: STRUCT = 1;
	*/
	DatatypeDescriptor_DatatypeKind$1[DatatypeDescriptor_DatatypeKind$1["STRUCT"] = 1] = "STRUCT";
	/**
	* @generated from protobuf enum value: ENUM = 2;
	*/
	DatatypeDescriptor_DatatypeKind$1[DatatypeDescriptor_DatatypeKind$1["ENUM"] = 2] = "ENUM";
	return DatatypeDescriptor_DatatypeKind$1;
}({});
/**
* @generated from protobuf enum sui.rpc.v2.OpenSignatureBody.Type
*/
let OpenSignatureBody_Type = /* @__PURE__ */ function(OpenSignatureBody_Type$1) {
	/**
	* @generated from protobuf enum value: TYPE_UNKNOWN = 0;
	*/
	OpenSignatureBody_Type$1[OpenSignatureBody_Type$1["TYPE_UNKNOWN"] = 0] = "TYPE_UNKNOWN";
	/**
	* @generated from protobuf enum value: ADDRESS = 1;
	*/
	OpenSignatureBody_Type$1[OpenSignatureBody_Type$1["ADDRESS"] = 1] = "ADDRESS";
	/**
	* @generated from protobuf enum value: BOOL = 2;
	*/
	OpenSignatureBody_Type$1[OpenSignatureBody_Type$1["BOOL"] = 2] = "BOOL";
	/**
	* @generated from protobuf enum value: U8 = 3;
	*/
	OpenSignatureBody_Type$1[OpenSignatureBody_Type$1["U8"] = 3] = "U8";
	/**
	* @generated from protobuf enum value: U16 = 4;
	*/
	OpenSignatureBody_Type$1[OpenSignatureBody_Type$1["U16"] = 4] = "U16";
	/**
	* @generated from protobuf enum value: U32 = 5;
	*/
	OpenSignatureBody_Type$1[OpenSignatureBody_Type$1["U32"] = 5] = "U32";
	/**
	* @generated from protobuf enum value: U64 = 6;
	*/
	OpenSignatureBody_Type$1[OpenSignatureBody_Type$1["U64"] = 6] = "U64";
	/**
	* @generated from protobuf enum value: U128 = 7;
	*/
	OpenSignatureBody_Type$1[OpenSignatureBody_Type$1["U128"] = 7] = "U128";
	/**
	* @generated from protobuf enum value: U256 = 8;
	*/
	OpenSignatureBody_Type$1[OpenSignatureBody_Type$1["U256"] = 8] = "U256";
	/**
	* @generated from protobuf enum value: VECTOR = 9;
	*/
	OpenSignatureBody_Type$1[OpenSignatureBody_Type$1["VECTOR"] = 9] = "VECTOR";
	/**
	* @generated from protobuf enum value: DATATYPE = 10;
	*/
	OpenSignatureBody_Type$1[OpenSignatureBody_Type$1["DATATYPE"] = 10] = "DATATYPE";
	/**
	* @generated from protobuf enum value: TYPE_PARAMETER = 11;
	*/
	OpenSignatureBody_Type$1[OpenSignatureBody_Type$1["TYPE_PARAMETER"] = 11] = "TYPE_PARAMETER";
	return OpenSignatureBody_Type$1;
}({});
/**
* @generated from protobuf enum sui.rpc.v2.FunctionDescriptor.Visibility
*/
let FunctionDescriptor_Visibility = /* @__PURE__ */ function(FunctionDescriptor_Visibility$1) {
	/**
	* @generated from protobuf enum value: VISIBILITY_UNKNOWN = 0;
	*/
	FunctionDescriptor_Visibility$1[FunctionDescriptor_Visibility$1["VISIBILITY_UNKNOWN"] = 0] = "VISIBILITY_UNKNOWN";
	/**
	* @generated from protobuf enum value: PRIVATE = 1;
	*/
	FunctionDescriptor_Visibility$1[FunctionDescriptor_Visibility$1["PRIVATE"] = 1] = "PRIVATE";
	/**
	* @generated from protobuf enum value: PUBLIC = 2;
	*/
	FunctionDescriptor_Visibility$1[FunctionDescriptor_Visibility$1["PUBLIC"] = 2] = "PUBLIC";
	/**
	* @generated from protobuf enum value: FRIEND = 3;
	*/
	FunctionDescriptor_Visibility$1[FunctionDescriptor_Visibility$1["FRIEND"] = 3] = "FRIEND";
	return FunctionDescriptor_Visibility$1;
}({});
/**
* @generated from protobuf enum sui.rpc.v2.OpenSignature.Reference
*/
let OpenSignature_Reference = /* @__PURE__ */ function(OpenSignature_Reference$1) {
	/**
	* @generated from protobuf enum value: REFERENCE_UNKNOWN = 0;
	*/
	OpenSignature_Reference$1[OpenSignature_Reference$1["REFERENCE_UNKNOWN"] = 0] = "REFERENCE_UNKNOWN";
	/**
	* @generated from protobuf enum value: IMMUTABLE = 1;
	*/
	OpenSignature_Reference$1[OpenSignature_Reference$1["IMMUTABLE"] = 1] = "IMMUTABLE";
	/**
	* @generated from protobuf enum value: MUTABLE = 2;
	*/
	OpenSignature_Reference$1[OpenSignature_Reference$1["MUTABLE"] = 2] = "MUTABLE";
	return OpenSignature_Reference$1;
}({});
/**
* An `Ability` classifies what operations are permitted for a given type
*
* @generated from protobuf enum sui.rpc.v2.Ability
*/
let Ability = /* @__PURE__ */ function(Ability$1) {
	/**
	* @generated from protobuf enum value: ABILITY_UNKNOWN = 0;
	*/
	Ability$1[Ability$1["ABILITY_UNKNOWN"] = 0] = "ABILITY_UNKNOWN";
	/**
	* Allows values of types with this ability to be copied
	*
	* @generated from protobuf enum value: COPY = 1;
	*/
	Ability$1[Ability$1["COPY"] = 1] = "COPY";
	/**
	* Allows values of types with this ability to be dropped.
	*
	* @generated from protobuf enum value: DROP = 2;
	*/
	Ability$1[Ability$1["DROP"] = 2] = "DROP";
	/**
	* Allows values of types with this ability to exist inside a struct in global storage
	*
	* @generated from protobuf enum value: STORE = 3;
	*/
	Ability$1[Ability$1["STORE"] = 3] = "STORE";
	/**
	* Allows the type to serve as a key for global storage operations
	*
	* @generated from protobuf enum value: KEY = 4;
	*/
	Ability$1[Ability$1["KEY"] = 4] = "KEY";
	return Ability$1;
}({});
var Package$Type = class extends MessageType {
	constructor() {
		super("sui.rpc.v2.Package", [
			{
				no: 1,
				name: "storage_id",
				kind: "scalar",
				opt: true,
				T: 9
			},
			{
				no: 2,
				name: "original_id",
				kind: "scalar",
				opt: true,
				T: 9
			},
			{
				no: 3,
				name: "version",
				kind: "scalar",
				opt: true,
				T: 4,
				L: 0
			},
			{
				no: 4,
				name: "modules",
				kind: "message",
				repeat: 1,
				T: () => Module
			},
			{
				no: 5,
				name: "type_origins",
				kind: "message",
				repeat: 1,
				T: () => TypeOrigin
			},
			{
				no: 6,
				name: "linkage",
				kind: "message",
				repeat: 1,
				T: () => Linkage
			}
		]);
	}
};
/**
* @generated MessageType for protobuf message sui.rpc.v2.Package
*/
const Package = new Package$Type();
var Module$Type = class extends MessageType {
	constructor() {
		super("sui.rpc.v2.Module", [
			{
				no: 1,
				name: "name",
				kind: "scalar",
				opt: true,
				T: 9
			},
			{
				no: 2,
				name: "contents",
				kind: "scalar",
				opt: true,
				T: 12
			},
			{
				no: 3,
				name: "datatypes",
				kind: "message",
				repeat: 1,
				T: () => DatatypeDescriptor
			},
			{
				no: 4,
				name: "functions",
				kind: "message",
				repeat: 1,
				T: () => FunctionDescriptor
			}
		]);
	}
};
/**
* @generated MessageType for protobuf message sui.rpc.v2.Module
*/
const Module = new Module$Type();
var DatatypeDescriptor$Type = class extends MessageType {
	constructor() {
		super("sui.rpc.v2.DatatypeDescriptor", [
			{
				no: 1,
				name: "type_name",
				kind: "scalar",
				opt: true,
				T: 9
			},
			{
				no: 2,
				name: "defining_id",
				kind: "scalar",
				opt: true,
				T: 9
			},
			{
				no: 3,
				name: "module",
				kind: "scalar",
				opt: true,
				T: 9
			},
			{
				no: 4,
				name: "name",
				kind: "scalar",
				opt: true,
				T: 9
			},
			{
				no: 5,
				name: "abilities",
				kind: "enum",
				repeat: 1,
				T: () => ["sui.rpc.v2.Ability", Ability]
			},
			{
				no: 6,
				name: "type_parameters",
				kind: "message",
				repeat: 1,
				T: () => TypeParameter
			},
			{
				no: 7,
				name: "kind",
				kind: "enum",
				opt: true,
				T: () => ["sui.rpc.v2.DatatypeDescriptor.DatatypeKind", DatatypeDescriptor_DatatypeKind]
			},
			{
				no: 8,
				name: "fields",
				kind: "message",
				repeat: 1,
				T: () => FieldDescriptor
			},
			{
				no: 9,
				name: "variants",
				kind: "message",
				repeat: 1,
				T: () => VariantDescriptor
			}
		]);
	}
};
/**
* @generated MessageType for protobuf message sui.rpc.v2.DatatypeDescriptor
*/
const DatatypeDescriptor = new DatatypeDescriptor$Type();
var TypeParameter$Type = class extends MessageType {
	constructor() {
		super("sui.rpc.v2.TypeParameter", [{
			no: 1,
			name: "constraints",
			kind: "enum",
			repeat: 1,
			T: () => ["sui.rpc.v2.Ability", Ability]
		}, {
			no: 2,
			name: "is_phantom",
			kind: "scalar",
			opt: true,
			T: 8
		}]);
	}
};
/**
* @generated MessageType for protobuf message sui.rpc.v2.TypeParameter
*/
const TypeParameter = new TypeParameter$Type();
var FieldDescriptor$Type = class extends MessageType {
	constructor() {
		super("sui.rpc.v2.FieldDescriptor", [
			{
				no: 1,
				name: "name",
				kind: "scalar",
				opt: true,
				T: 9
			},
			{
				no: 2,
				name: "position",
				kind: "scalar",
				opt: true,
				T: 13
			},
			{
				no: 3,
				name: "type",
				kind: "message",
				T: () => OpenSignatureBody
			}
		]);
	}
};
/**
* @generated MessageType for protobuf message sui.rpc.v2.FieldDescriptor
*/
const FieldDescriptor = new FieldDescriptor$Type();
var VariantDescriptor$Type = class extends MessageType {
	constructor() {
		super("sui.rpc.v2.VariantDescriptor", [
			{
				no: 1,
				name: "name",
				kind: "scalar",
				opt: true,
				T: 9
			},
			{
				no: 2,
				name: "position",
				kind: "scalar",
				opt: true,
				T: 13
			},
			{
				no: 3,
				name: "fields",
				kind: "message",
				repeat: 1,
				T: () => FieldDescriptor
			}
		]);
	}
};
/**
* @generated MessageType for protobuf message sui.rpc.v2.VariantDescriptor
*/
const VariantDescriptor = new VariantDescriptor$Type();
var OpenSignatureBody$Type = class extends MessageType {
	constructor() {
		super("sui.rpc.v2.OpenSignatureBody", [
			{
				no: 1,
				name: "type",
				kind: "enum",
				opt: true,
				T: () => ["sui.rpc.v2.OpenSignatureBody.Type", OpenSignatureBody_Type]
			},
			{
				no: 2,
				name: "type_name",
				kind: "scalar",
				opt: true,
				T: 9
			},
			{
				no: 3,
				name: "type_parameter_instantiation",
				kind: "message",
				repeat: 1,
				T: () => OpenSignatureBody
			},
			{
				no: 4,
				name: "type_parameter",
				kind: "scalar",
				opt: true,
				T: 13
			}
		]);
	}
};
/**
* @generated MessageType for protobuf message sui.rpc.v2.OpenSignatureBody
*/
const OpenSignatureBody = new OpenSignatureBody$Type();
var FunctionDescriptor$Type = class extends MessageType {
	constructor() {
		super("sui.rpc.v2.FunctionDescriptor", [
			{
				no: 1,
				name: "name",
				kind: "scalar",
				opt: true,
				T: 9
			},
			{
				no: 5,
				name: "visibility",
				kind: "enum",
				opt: true,
				T: () => ["sui.rpc.v2.FunctionDescriptor.Visibility", FunctionDescriptor_Visibility]
			},
			{
				no: 6,
				name: "is_entry",
				kind: "scalar",
				opt: true,
				T: 8
			},
			{
				no: 7,
				name: "type_parameters",
				kind: "message",
				repeat: 1,
				T: () => TypeParameter
			},
			{
				no: 8,
				name: "parameters",
				kind: "message",
				repeat: 1,
				T: () => OpenSignature
			},
			{
				no: 9,
				name: "returns",
				kind: "message",
				repeat: 1,
				T: () => OpenSignature
			}
		]);
	}
};
/**
* @generated MessageType for protobuf message sui.rpc.v2.FunctionDescriptor
*/
const FunctionDescriptor = new FunctionDescriptor$Type();
var OpenSignature$Type = class extends MessageType {
	constructor() {
		super("sui.rpc.v2.OpenSignature", [{
			no: 1,
			name: "reference",
			kind: "enum",
			opt: true,
			T: () => ["sui.rpc.v2.OpenSignature.Reference", OpenSignature_Reference]
		}, {
			no: 2,
			name: "body",
			kind: "message",
			T: () => OpenSignatureBody
		}]);
	}
};
/**
* @generated MessageType for protobuf message sui.rpc.v2.OpenSignature
*/
const OpenSignature = new OpenSignature$Type();
var TypeOrigin$Type = class extends MessageType {
	constructor() {
		super("sui.rpc.v2.TypeOrigin", [
			{
				no: 1,
				name: "module_name",
				kind: "scalar",
				opt: true,
				T: 9
			},
			{
				no: 2,
				name: "datatype_name",
				kind: "scalar",
				opt: true,
				T: 9
			},
			{
				no: 3,
				name: "package_id",
				kind: "scalar",
				opt: true,
				T: 9
			}
		]);
	}
};
/**
* @generated MessageType for protobuf message sui.rpc.v2.TypeOrigin
*/
const TypeOrigin = new TypeOrigin$Type();
var Linkage$Type = class extends MessageType {
	constructor() {
		super("sui.rpc.v2.Linkage", [
			{
				no: 1,
				name: "original_id",
				kind: "scalar",
				opt: true,
				T: 9
			},
			{
				no: 2,
				name: "upgraded_id",
				kind: "scalar",
				opt: true,
				T: 9
			},
			{
				no: 3,
				name: "upgraded_version",
				kind: "scalar",
				opt: true,
				T: 4,
				L: 0
			}
		]);
	}
};
/**
* @generated MessageType for protobuf message sui.rpc.v2.Linkage
*/
const Linkage = new Linkage$Type();

//#endregion
export { Ability, DatatypeDescriptor, DatatypeDescriptor_DatatypeKind, FieldDescriptor, FunctionDescriptor, FunctionDescriptor_Visibility, Linkage, Module, OpenSignature, OpenSignatureBody, OpenSignatureBody_Type, OpenSignature_Reference, Package, TypeOrigin, TypeParameter, VariantDescriptor };
