// @mysten/sui/dist/grpc/proto/sui/rpc/v2/owner.mjs
import { MessageType } from "@protobuf-ts/runtime";

//#region src/grpc/proto/sui/rpc/v2/owner.ts
/**
* @generated from protobuf enum sui.rpc.v2.Owner.OwnerKind
*/
let Owner_OwnerKind = /* @__PURE__ */ function(Owner_OwnerKind$1) {
	/**
	* @generated from protobuf enum value: OWNER_KIND_UNKNOWN = 0;
	*/
	Owner_OwnerKind$1[Owner_OwnerKind$1["OWNER_KIND_UNKNOWN"] = 0] = "OWNER_KIND_UNKNOWN";
	/**
	* @generated from protobuf enum value: ADDRESS = 1;
	*/
	Owner_OwnerKind$1[Owner_OwnerKind$1["ADDRESS"] = 1] = "ADDRESS";
	/**
	* @generated from protobuf enum value: OBJECT = 2;
	*/
	Owner_OwnerKind$1[Owner_OwnerKind$1["OBJECT"] = 2] = "OBJECT";
	/**
	* @generated from protobuf enum value: SHARED = 3;
	*/
	Owner_OwnerKind$1[Owner_OwnerKind$1["SHARED"] = 3] = "SHARED";
	/**
	* @generated from protobuf enum value: IMMUTABLE = 4;
	*/
	Owner_OwnerKind$1[Owner_OwnerKind$1["IMMUTABLE"] = 4] = "IMMUTABLE";
	/**
	* @generated from protobuf enum value: CONSENSUS_ADDRESS = 5;
	*/
	Owner_OwnerKind$1[Owner_OwnerKind$1["CONSENSUS_ADDRESS"] = 5] = "CONSENSUS_ADDRESS";
	return Owner_OwnerKind$1;
}({});
var Owner$Type = class extends MessageType {
	constructor() {
		super("sui.rpc.v2.Owner", [
			{
				no: 1,
				name: "kind",
				kind: "enum",
				opt: true,
				T: () => ["sui.rpc.v2.Owner.OwnerKind", Owner_OwnerKind]
			},
			{
				no: 2,
				name: "address",
				kind: "scalar",
				opt: true,
				T: 9
			},
			{
				no: 3,
				name: "version",
				kind: "scalar",
				opt: true,
				T: 4,
				L: 0
			}
		]);
	}
};
/**
* @generated MessageType for protobuf message sui.rpc.v2.Owner
*/
const Owner = new Owner$Type();

//#endregion
export { Owner, Owner_OwnerKind };
