// @mysten/sui/dist/grpc/proto/sui/rpc/v2/signature_scheme.mjs
//#region src/grpc/proto/sui/rpc/v2/signature_scheme.ts
/**
* Flag use to disambiguate the signature schemes supported by Sui.
*
* Note: the enum values defined by this proto message exactly match their
* expected BCS serialized values when serialized as a u8. See
* [enum.SignatureScheme](https://mystenlabs.github.io/sui-rust-sdk/sui_sdk_types/enum.SignatureScheme.html)
* for more information about signature schemes.
*
* @generated from protobuf enum sui.rpc.v2.SignatureScheme
*/
let SignatureScheme = /* @__PURE__ */ function(SignatureScheme$1) {
	/**
	* @generated from protobuf enum value: ED25519 = 0;
	*/
	SignatureScheme$1[SignatureScheme$1["ED25519"] = 0] = "ED25519";
	/**
	* @generated from protobuf enum value: SECP256K1 = 1;
	*/
	SignatureScheme$1[SignatureScheme$1["SECP256K1"] = 1] = "SECP256K1";
	/**
	* @generated from protobuf enum value: SECP256R1 = 2;
	*/
	SignatureScheme$1[SignatureScheme$1["SECP256R1"] = 2] = "SECP256R1";
	/**
	* @generated from protobuf enum value: MULTISIG = 3;
	*/
	SignatureScheme$1[SignatureScheme$1["MULTISIG"] = 3] = "MULTISIG";
	/**
	* @generated from protobuf enum value: BLS12381 = 4;
	*/
	SignatureScheme$1[SignatureScheme$1["BLS12381"] = 4] = "BLS12381";
	/**
	* @generated from protobuf enum value: ZKLOGIN = 5;
	*/
	SignatureScheme$1[SignatureScheme$1["ZKLOGIN"] = 5] = "ZKLOGIN";
	/**
	* @generated from protobuf enum value: PASSKEY = 6;
	*/
	SignatureScheme$1[SignatureScheme$1["PASSKEY"] = 6] = "PASSKEY";
	return SignatureScheme$1;
}({});

//#endregion
export { SignatureScheme };
