// @evefrontier/dapp-kit/utils/index.ts
// Transform utilities
export {
  parseStatus,
  getAssemblyType,
  getObjectId,
  getRegistryAddress,
} from "./mapping";
export { getEnergyUsageForType, getFuelEfficiencyForType } from "./config";
export {
  getAdjustedBurnRate,
  type AdjustedBurnRate,
} from "./burnRate";
export { parseCharacterFromJson } from "./character";
export {
  getCharacterOwnedObjectsJson,
  getCharacterOwnedObjects,
} from "./characterOwnedObjects";
export { transformToAssembly, transformToCharacter } from "./transforms";
export type { TransformOptions } from "./transforms";

// Datahub utilities
export { getDatahubGameInfo } from "./datahub";

// General utilities
export {
  assertAssemblyType,
  abbreviateAddress,
  clickToCopy,
  findOwnerByAddress,
  formatM3,
  formatDuration,
  getCommonItems,
  getDappUrl,
  getEnv,
  getTxUrl,
  getVolumeM3,
  isOwner,
  parseURL,
  removeTrailingZeros,
} from "./utils";

// Constants
export * from "./constants";

// Error handling
export { ERRORS, ERROR_MESSAGES, parseErrorFromMessage } from "./errors";
export type { ErrorType } from "./errors";
